### PR TITLE
Copy variadic arguments to buffer to avoid horrific varargs hacks

### PR DESCRIPTION
**Pull request by yebblies**
The current implementation of `doFormat` forges a `va_list` so it can recurse to print out array and AA members.  This was straightforward on X86, but the complexity of X86_64 varargs abis means that it has to jump through some rather horrific and non-portable hoops to get this to work.

This replacement implementation instead copies all variadic arguments to a buffer, being careful to preserve required alignment.  This allows the arguments to be iterated through with a simple pointer on all platforms.

The copy to the buffer naturally comes with a slight performance penalty, but given the simplification of the code, the improved portability, and the fact that `doFormat`'s only remaining caller is `std.stream`, I think it's worth that cost.

@andralex @ibuclaw

### CODE REVIEW

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by andralex** [comment]
rox

**Comment by Geod24** [comment]
Nice improvement, but shouldn't we deprecate / remove this instead (Along with std.socket's functions that are using it) ?

**Comment by yebblies** [comment]
> Nice improvement, but shouldn't we deprecate / remove this instead (Along with std.socket's functions that are using it) ?

Sure, but I have no intention of having my varargs work get bogged down in the discussion of std.stream's replacement.

**Comment by yebblies** [comment]
Now I guess I need to fix `va_save`.

File: std/format.d [hunk 1]
```diff
@@ -5257,13 +5257,54 @@ void main()
...
+        bufUsed -= (cast(size_t)argBuffer + bufUsed) & (ti.talign - 1);
+        auto pos = bufUsed;
+        // Align to next word boundary
+        bufUsed += ti.tsize + size_t.sizeof - 1;
+        bufUsed -= (cast(size_t)argBuffer + bufUsed) & (size_t.sizeof - 1);
+        // Resize buffer if necessary
+        while (bufUsed > bufLength)
+        {
+            bufLength *= 2;
+            argBuffer = realloc(argBuffer, bufLength);
```
Comment (ZombineDev) [inline std/format.d]: @yebblies I think `realloc` should be placed after the loop (if I am not missing something).

Comment (yebblies) [inline std/format.d]: The loop is highly unlikely to execute more than once.

Comment (yebblies) [inline std/format.d]: If it executes at all.

**Comment by ibuclaw** [comment]
@klickverbot - does LDC patch core.stdc.stdarg to work with more targets than just ARM, x86 and x86_64?  Referring specifically to ``void va_arg()(ref va_list apx, TypeInfo ti, void* parmn)``

**Comment by JakobOvrum** [comment]
Did this break Phobos compilation? I get:
```
..\..\src\druntime\import\core\stdc\stdarg.d(67): Error: cannot implicitly convert expression (cast(void*)(cast(uint)p + (tsize + 4u - 1u & 4294967292u))) of type void* to char*
std\format.d(5288): Error: template instance core.stdc.stdarg.va_arg!() error instantiating
std\stream.d(1233):        instantiated from here: doFormat!()
```
Using `master` on all repositories as of the time of posting.

**Comment by ibuclaw** [comment]
The autotest seems happy enough....

https://auto-tester.puremagic.com/pulls.ghtml?projectid=1

**Comment by yebblies** [comment]
> Using master on all repositories as of the time of posting.

I would guess your druntime import folder is stale, and you need to run 'make import".

**Comment by yebblies** [comment]
Specifically you're missing the changes in this PR: https://github.com/D-Programming-Language/druntime/pull/1133

**Comment by JakobOvrum** [comment]
Oops, I had changed a line in druntime's makefile to *not* try to compile `minit.asm`, which was preventing my pull script from pulling the latest druntime changes. Thanks!

(... it's quite annoying that the `clean` target deletes `minit.obj`)

**Comment by yebblies** [comment]
> (... it's quite annoying that the clean target deletes minit.obj)

But it doesn't.

**Comment by yebblies** [comment]
@ibuclaw I've given up on fixing va_list's type on 64-bit for now.  Everything I do just ends up iceing the backend.

**Comment by ibuclaw** [comment]
@yebblies - that's pretty much what I found when doing the vice versa with GDC (changing the va_list type to be wrapped around ``void*``).

Sad times.  But glad a start has been made.

**Comment by ibuclaw** [comment]
@yebblies by the way - I'm currently stuck on 2.065's implementation of cppmangle, as someone decided to make static array's an ICE.

**Comment by yebblies** [comment]
Static arrays don't exactly have a mapping to C++ types... are you planning to mangle them like C++ arrays.  It's a pity D's static arrays don't work like D1's.

I'm thinking I might have a go at changing `va_copy` to alloca some memory, that would work most of the time, at least better than the current situation.

**Comment by ibuclaw** [comment]
Mangled and passed as C arrays.  Otherwise functions such as vaprintf will not work.  Current workaround is to rewrite the va parameters as ref, but the compiler only knows how to deal with gcc.builtins and core.stdc packages.
This is really only to deal with va_list, and not static arrays in general.

**Comment by yebblies** [comment]
I guess that works, although it's a shame to go through all this for one stupid abi.

**Comment by ibuclaw** [comment]
A couple of stupid abi's (x86_64 isn't the only one :)

**Comment by ibuclaw** [comment]
Also, a couple more use a struct for va_list, and so pointer semantics cannot be assumed either.

**Comment by yebblies** [comment]
Struct `va_list` is fine, because we can just alias it.  This would be a lot easier if they had defined separate `va_list_stack` and `va_list_arg` types.

Anyway, changing `va_copy` to `void va_copy(out va_list dest, va_list src, void* storage = alloca(__va_list_tag.sizeof))` seems to work, which will finally fix the crash I'm having in dmd.  It obviously won't work in all situations, but it should do for the vast majority.

**Comment by yebblies** [comment]
I've opened a bug for va_copy not working.

https://issues.dlang.org/show_bug.cgi?id=14131

**Comment by ibuclaw** [comment]
Don't you mean ``va_list_arg`` and ``va_list_ref``?  I'll have a look...

**Comment by yebblies** [comment]
> Don't you mean va_list_arg and va_list_ref? I'll have a look...

I mean if C had defined it in terms of types D actually had then it would be a lot easier for us.

**Pull request merged**